gateway/ln-gateway/src/events.rs

```diff
@@ -22,4 +32,288 @@ pub const ALL_GATEWAY_EVENTS: [EventKind; 11] = [
+            incoming_fees: incoming_success_stats.iter().map(|(_, f)| *f).collect(),
+            outgoing_latencies_failure: outgoing_failure_stats,
+            incoming_latencies_failure: incoming_failure_stats,
+        };
+        events.sort();
+        events
+    }
+
+    /// Combines this `StructuredPaymentEvents` with the `other`
+    /// `StructuredPaymentEvents` by appending all of the internal vectors.
+    pub fn combine(&mut self, other: &mut StructuredPaymentEvents) {
+        self.outgoing_latencies
+            .append(&mut other.outgoing_latencies);
+        self.incoming_latencies
+            .append(&mut other.incoming_latencies);
+        self.outgoing_fees.append(&mut other.outgoing_fees);
+        self.incoming_fees.append(&mut other.incoming_fees);
+        self.outgoing_latencies_failure
+            .append(&mut other.outgoing_latencies_failure);
+        self.incoming_latencies_failure
+            .append(&mut other.incoming_latencies_failure);
+        self.sort();
+    }
+
+    /// Sorts this `StructuredPaymentEvents` by sorting all of the internal
+    /// vectors.
+    fn sort(&mut self) {
+        self.outgoing_latencies.sort_unstable();
+        self.incoming_latencies.sort_unstable();
+        self.outgoing_fees.sort_unstable();
+        self.incoming_fees.sort_unstable();
+        self.outgoing_latencies_failure.sort_unstable();
+        self.incoming_latencies_failure.sort_unstable();
+    }
+
+    /// Computes a `PaymentSummaryResponse` that can display at a glance
+    /// statistics about the outgoing and incoming payments.
+    pub fn payment_summary_response(&self) -> PaymentSummaryResponse {
+        PaymentSummaryResponse {
+            outgoing: Self::compute_payment_stats(
+                &self.outgoing_latencies,
+                &self.outgoing_fees,
+                &self.outgoing_latencies_failure,
+            ),
+            incoming: Self::compute_payment_stats(
+                &self.incoming_latencies,
+                &self.incoming_fees,
+                &self.incoming_latencies_failure,
+            ),
+        }
+    }
+
+    /// Computes the payment statistics for the given input data.
+    fn compute_payment_stats(
+        latencies: &[u64],
+        fees: &[Amount],
+        latencies_failure: &[u64],
+    ) -> PaymentStats {
+        PaymentStats {
+            average_latency_micros: average(latencies),
+            median_latency_micros: median(latencies),
+            total_fees: Amount::from_msats(fees.iter().map(|a| a.msats).sum()),
+            total_success: latencies.len(),
+            total_failure: latencies_failure.len(),
+        }
+    }
+}
+
+/// Computes the average of the given `u64` slice.
+fn average(data: &[u64]) -> u64 {
+    let sum: u64 = data.iter().sum();
+    if data.is_empty() {
+        0
+    } else {
+        sum / data.len() as u64
+    }
+}
+
+/// Computes the median of the given `u64` slice.
```

Comment:
*sorted slice, right?

We have a median in `modules/fedimint-wallet-server/src/lib.rs` maybe let's put in `core` somewhere?

Reply:
Why does Rust not have a median from sorted array in a stdlib?

Reply:
Done